### ISSUE TITLE
tf.data.Iterator has_next(), next(), feature request

### ISSUE THREAD

**Issue by jackshi0912** [tensorflow/tensorflow#20986]
Request:
has_next()  // opt that check if next data is present at graph execution time without resorting to python try/except
next()  // op that advance the iterator by one at graph execution time

Reason:
has_next():
Iterator seems to be designed for supervised learning. Data is fed from one end, OutOfRangeError occurs when all the data has been used and training stops (if repeat is not set).

In reinforcement learning, the end of episode (last example in file) can be a significant source of information during graph execution time. Catch the exception and run an additional ad-hoc training cycle for the last example is aweful.

next():
In RL, the data can be sequential and has specific time stamps. During graph execution time, the environment written in tensorflow should be able to advance the Iterator using tf.while_loop with iterator.next() until some condition is met 

for example, advance Iterator in a tf.while_loop until time stamp associated with that example is greater than some internal state (variable).

Perhaps the get_next() method should be parameter such that the node can be optionally required to re-executed after every visit. or just have a next() method and let us figure out the appropriate dependency control.

Have I written custom code: Yes
OS Platform and Distribution: Ubuntu 16.04 LTS
TensorFlow installed from: pip
TensorFlow version: 1.8
Bazel version: 0.15.2
CUDA/cuDNN version: Cuda 9.0, cuDNN v7
GPU model and memory: GTX 1070 8GB
Exact command to reproduce: N/A
Mobile device: N/A

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by mrry** [comment]
I recently added `tf.contrib.data.get_next_as_optional()` (9811b72df7347a4ec2ee884a2165775340d70e75), which makes it possible to consume values from an iterator in a `tf.while_loop()` without raising an exception at the end of the sequence. This should make it possible to implement your RL use case.

As I commented on #20255, we have no plans to add `Iterator.has_next()`, but a combination of `tf.contrib.data.get_next_as_optional()` and local variables should suffice to implement your use case.
